MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java

```diff
@@ -0,0 +1,243 @@
+package com.google.android.libraries.motionphotoreader;
+
+import android.graphics.SurfaceTexture;
+import android.opengl.GLES11;
+import android.opengl.GLES11Ext;
+import android.opengl.GLES20;
+import android.opengl.Matrix;
+import android.util.Log;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.FloatBuffer;
+
+import javax.microedition.khronos.egl.EGL;
+import javax.microedition.khronos.opengles.GL;
+
+/**
+ * Renders frames from a MediaCodec decoder (obtained from a surface texture attached to an OpenGL
+ * texture) onto an EGL surface.
+ */
+public class TextureRender {
+
+    private static final String TAG = "TextureRender";
+
+    private static final int FLOAT_SIZE_BYTES = 4; // number of bytes for a float
+    private static final int TRIANGLE_VERTICES_DATA_POS_OFFSET = 0;
+    private static final int TRIANGLE_VERTICES_DATA_STRIDE_BYTES = 5 * FLOAT_SIZE_BYTES; // 5 due to five coordinate float values (x,y,z,u,v)
+    private static final int TRIANGLE_VERTICES_DATA_UV_OFFSET = 3; // texture coords begin 3 offset units from array start (skip the position coordinates)
+
+    private static final String VERTEX_SHADER =
+            "uniform mat4 u_Matrix;\n" +
+            "attribute vec4 a_Position;\n" +
+            "attribute vec2 a_TextureCoordinates;\n" +
+            "varying vec2 v_TextureCoordinates;\n" +
+            "void main() {\n" +
+            "  v_TextureCoordinates = a_TextureCoordinates;\n" +
+            "  gl_Position = u_Matrix * a_Position;\n" +
+            "}";
+
+    private static final String FRAGMENT_SHADER =
+            "precision mediump float;\n" +
+            "uniform sampler2D u_TextureUnit;\n" +
+            "varying vec2 v_TextureCoordinates;\n" +
+            "void main() {\n" +
+            "  gl_FragColor = texture2D(u_TextureUnit, v_TextureCoordinates);\n" +
+            "}";
+
+    // A single plane comprised of two triangles to hold the image texture
+    private final float[] triangleVerticesData = {
+            // positions (x,y,z)    // texture coordinates (u,v)
+            -1.0f, -1.0f, 0.0f,     0.0f, 0.0f,  // bottom left
+             1.0f, -1.0f, 0.0f,     1,0f, 0.0f,  // bottom right
```

Comment:
There is a very sneaky mistake here: this line contains SIX numbers and not FIVE. That caused a bunch of issues in your triangles.